docs/develop/python/message-passing.mdx

```diff
@@ -413,131 +499,119 @@ class MyWorkflow:
+  This raises a [`temporalio.client.WorkflowUpdateRPCTimeoutOrCancelledError`](https://python.temporal.io/temporalio.client.WorkflowUpdateRPCTimeoutOrCancelledError.html) exception.
 
-### Problems when sending an Update
+- **Update failed**: You'll receive a [`temporalio.client.WorkflowUpdateFailedError`](https://python.temporal.io/temporalio.client.WorkflowUpdateFailedError.html) exception.
+  There are two ways this can happen:
 
-1. **There is no Workflow Worker polling the Task Queue**
+    - The Update was rejected by an Update validator defined in the Workflow alongside the Update handler.
 
-    Your request will be retried by the SDK Client indefinitely.
-    You can use [`asyncio.timeout`](https://docs.python.org/3/library/asyncio-task.html#timeouts) to impose a timeout; this will cause it to raise [`temporalio.client.WorkflowUpdateRPCTimeoutOrCancelledError`](https://python.temporal.io/temporalio.client.WorkflowUpdateRPCTimeoutOrCancelledError.html).
+    - The Update failed after having been accepted.
 
-1. **Update failed.**
+    Update failures are like [Workflow failures](/references/failures#errors-in-workflows).
+    Issues that cause a Workflow failure in the main method also cause Update failures in the Update handler.
+    These might include:
 
-    You'll get [`temporalio.client.WorkflowUpdateFailedError`](https://python.temporal.io/temporalio.client.WorkflowUpdateFailedError.html). There are two ways this can happen:
+      - A failed Child Workflow
+      - A failed Activity (if the Activity retries have been set to a finite number)
+      - The Workflow author raising `ApplicationFailure`
+      - Any error listed in [workflow_failure_exception_types](https://python.temporal.io/temporalio.worker.Worker.html) (empty by default)
 
-    1. The Update was [rejected](#handle-update) by an Update validator (defined in the Workflow alongside the Update handler).
+- **The handler caused the Workflow Task to fail**:
+  A [Workflow Task Failure](/references/failures#errors-in-workflows) causes the server to retry Workflow Tasks indefinitely. What happens to your Update request depends on its stage:
+    - If the request hasn't been accepted by the server, you receive a `FAILED_PRECONDITION` [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) exception.
+    - If the request has been accepted, it is durable.
+      Once the Workflow is healthy again after a code deploy, use an [`UpdateHandle`](https://python.temporal.io/temporalio.client.WorkflowUpdateHandle.html) to fetch the Update result.
 
-    1. The Update failed after having been accepted.
+- **The Workflow finished while the Update handler execution was in progress**:
+    You'll receive a [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) exception with a "workflow execution already completed" message.
```

Comment:
Hey @fairlydurable one more tiny change here -- it's a loose end I left and didn't get aorund to. Basically we weren't giving the RPC status code here (up to you whether we mention the error message)

```suggestion
You'll receive a [`temporalio.service.RPCError`](https://python.temporal.io/temporalio.service.RPCError.html) exception with a `status` attribute of [`RPCStatusCode`](https://python.temporal.io/temporalio.service.RPCStatusCode.html) `NOT_FOUND`.
```

Reply:
Adopted.